### ISSUE TITLE
TS incorrectly flags a side effect-only import as TS5097 (An import path can only end with a '.ts' extension...) in `--watch` mode

### ISSUE THREAD

**Issue by Aarononomous** [microsoft/TypeScript#58791]
### 🔎 Search Terms

"TS5097", "allowImportingTsExtensions", "noEmit", "watch", "side effect", "import"

### 🕗 Version & Regression Information

- This changed in commit or PR 89e928e8b41368bb0888c6d399854c53a2dbf3d4

### 💻 Code

When importing a `.ts` file strictly for side effects,

```ts
import './a.ts';
```

I expect that this will be accepted by the typechecker. In fact, even if a.ts didn't exist, this would still be valid TS, see https://github.com/microsoft/TypeScript/issues/50394#issuecomment-1381014487 for a motivating example.

Typechecking with `--noEmit` validates this without error.

Typechecking with `--watch` gives me `error TS5097: An import path can only end with a '.ts' extension when 'allowImportingTsExtensions' is enabled`.

**a.ts**
```ts
console.log("I'm a side effect!");
```

**b.ts**
```ts
import './a.ts';
```

**tsconfig.json**
```ts
{}
```

### 🙁 Actual behavior

```zsh
% npx tsc --noEmit # no output. Valid!
% npx tsc --watch
[5:54:03 PM] Starting compilation in watch mode...

b.ts:1:8 - error TS5097: An import path can only end with a '.ts' extension when 'allowImportingTsExtensions' is enabled.

1 import './a.ts';
         ~~~~~~~~

[5:54:04 PM] Found 1 error. Watching for file changes.
```

I shouldn't see a new error only in `--watch` mode.

### 🙂 Expected behavior

```zsh
% npx tsc --noEmit # no output. Valid.
% npx tsc --watch
[5:54:37 PM] File change detected. Starting incremental compilation...

[5:54:37 PM] Found 0 errors. Watching for file changes.
```

### Additional information about the issue

**Workaround:** `// @ts-ignore` before the import.

**Comment by Andarist** [comment]
accidentally-ish one of my recent PRs fixes this: https://github.com/microsoft/TypeScript/pull/58668

**Comment by fatcerberus** [comment]
@Andarist And you wanted to make this an error across the board... tsk tsk :wink:

**Comment by Andarist** [comment]
I didn't say I'm happy I fixed this, did I? :trollface:

**Comment by Andarist** [comment]
I really wonder why @Aarononomous is doing this in his code though. What's the reason behind importing a side-effectful `.ts` *without* enabling `allowImportingTsExtensions`? 🤔

**Comment by Aarononomous** [comment]
> I really wonder why @Aarononomous is doing this in his code though. What's the reason behind importing a side-effectful `.ts` _without_ enabling `allowImportingTsExtensions`? 🤔

I'm working in a multitenant codebase. We put platform-specific code in `*.platformName.ts` files, adding the platform name as an extension. Then we have a webpack resolver configured to look for platform files preferentially,

```ts
resolve: {
  extensions: [
    ...
    `.${CONFIG.platform}.ts`,
    `.${CONFIG.platform}.tsx`,
    '.ts',
    '.tsx',
    ...
```

so that when we build for, say, platformX and see `import './a'`, we first look for a file at `./a.platformX.ts`, then `./a.ts`. This allows us to write all of our platform-specific code without runtime tests or compiler macros or whatever, neatly encapsulated in its own file.

So not only do we not want to allow ts extensions, we _can't_! Otherwise our whole multitenant resolver behavior falls apart.

The exception to this is _in the multitenant files themselves_, where we do sometimes want to import common code from the extensionless base files, e.g.

**a.platformX.ts**
```ts
// @ts-expect-error
// We need the full filename with extension here: otherwise there's a circular import
import { module as baseModule } from './a.ts';

// do something with baseModule...

export const module;
```

The `'allowImportingTsExtensions'` error is expected when we import with a full filename. It's just when we only import for side effects from the base file that we see this inconsistency in `--watch` mode.

So the code example above is really more like

**a.ts**
```ts
window.__CONFIG__ = { ... };
```

**a.platformX.ts**
```ts
import './a.ts';

window.__CONFIG__.platformX = { ... };
```

**Comment by Andarist** [comment]
The use case makes sense but I think you could introduce some other convention for those base modules to avoid this problem altogether.

**Comment by Aarononomous** [comment]
@Andarist, what do you suggest?

If I put my side-effectful code in, say, **a.base.ts**, I'd still be calling it in **a.ts**, **a.platformX.ts**, etc. with

```ts
import 'a.base.ts';
```
